src/coreclr/jit/optimizer.cpp

```diff
@@ -2750,6 +2733,13 @@ void Compiler::optRedirectBlock(BasicBlock* blk, BlockToBlockMap* redirectMap, R
             break;
 
         case BBJ_ALWAYS:
+            // Fall-through successors are assumed correct and are not modified
```

Comment:
Is this new logic really necessary?

Reply:
I think so. That comment is copied from the notes in the doc comment for `Compiler::optRedirectBlock`; I added this check in to emulate the no-op behavior it previously had for `BBJ_NONE`. If I remove it, we hit `assert(h->HasJumpTo(t) || !h->KindIs(BBJ_ALWAYS))` in `Compiler::optCanonicalizeLoopCore`.

Reply:
The logic seems wrong with this added code here -- I would expect `optRedirectBlock` to always redirect a `BBJ_ALWAYS` based on the map. The behavior now doesn't match the documentation. Maybe some update to `redirectMap` is needed somewhere?

Reply:
To preserve the original behavior of this method, which was to not redirect `BBJ_NONE`, would it be ok to use `BBF_NONE_QUIRK` here instead to determine if we should redirect a `BBJ_ALWAYS`? This seems to work locally. e.g:
```
if (blk->JumpsToNext() && ((blk->bbFlags & BBF_NONE_QUIRK) != 0)) // Functionally equivalent to BBJ_NONE
```

Reply:
It would be better, even though I generally think quirks should not affect behavior in this way. It seems like there is some form of bug here around how the redirection happens or how the map is constructed.